### ISSUE TITLE
Reconsider `opam build`

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/opam#2882]
I just want to surface up the comments I made [here](https://github.com/ocaml/platform-blog/pull/42#discussion_r104774661) about `opam build`. 

I think the command, if kept at all, should at least be renamed as it is not a command that `builds` but installs. 

It seems to me that most if its functionality should actually be part of `opam install`, most of the options of `install` are actually relevant to `opam build` and vice-versa and more importantly you might want to use the functionality of `opam build` for a package that exists in a repo (e.g. redirect the install of the package to a directory of your choice). 

The whole `autoinit` business seems rather ad-hoc and limited (e.g. which repo will it use ?) and the actual use cases seem a bit unclear,  see [these comments](https://github.com/ocaml/platform-blog/pull/42#discussion_r104776288).

**Comment by AltGr** [comment]
About making the functionality available to `opam install`, I see a few parts that can be separated and are more generally useful:
- Use opam files from CWD. `opam pin <dir>` already does that. If it is useful to some to have the functionality available without requiring pinning,  we could have an `opam install` option `--packages-at <dir>` to use the package definitions directly and temporarily.
- Copy the package's installed files to a different destdir: `opam install foo --copy-to-destdir <dir>` could be added. Note that this could work for already installed packages without needing to recompile them.

The auto-init feature is more debatable:
- Running `opam init`: instead of stopping when `~/.opam` does not exist, we could add a message and propose "`Do the initialisation now ? [Y/n]`". I generally prefer to keep the commands separate, and this does only slightly improve the workflow compared to just displaying `Please run "opam init" first` as we do now.
- Creating a local switch in a single step: this reaches a bit further than the commands above, but it could be a new option to  `opam install` as well. Either propose to create the switch on e.g. `opam install --switch .`, or add something like `opam install --here`

**Comment by rgrinberg** [comment]
I agree that build is confusing and install is more appropriate. That being said, if `$ opam build` actually builds the local package, there's no reason it shouldn't run install for you if it sees that's missing. See how the stack tool works.


> opam install --switch ., or add something like opam install --here

`--here` or `--local` would be nice. To contrast with other package managers, npm for example where local is the default and you must specify `--global` to get the behavior we currently default to in opam.

**Comment by dbuenzli** [comment]
> That being said, if $ opam build actually builds the local package, there's no reason it shouldn't run install for you if it sees that's missing.

As I mentioned in the comments on the blog post, it took us a few `opam` versions before being able to make a distinction between `build:` and `install:` fields, I think it's better if we keep our concepts and vocabulary well disciplined. 

This `build` command is full of ambiguities (e.g. where is the local package built in the `cwd` or in `opam`'s build dir ?) and seems a bad idea altoghether: if you don't extract the install via `--install-dir` it will install in your switch and will *not* recompile if dependents upgrade, hello `make inconsistent assumptions about interface` messages. 

I really don't see the point of it, we have a well disciplined, correct and traceable way of doing this via `pins` there no need to add another half broken one for usage scenarios that so far are totally unconvincing.

**Comment by rgrinberg** [comment]
> I think it's better if we keep our concepts and vocabulary well disciplined.

Fine. Although I don't think convenience necessarily muddies anything. While inconvenience makes it hard to do the correct thing. Regardless of how well designed and orthogonal it is. 

> I really don't see the point of it, we have a well disciplined, correct and traceable way of doing this via pins there no need to add another half broken one for usage scenarios that so far are totally unconvincing.

I don't think pins solve my problems at all. What I want (or maybe what think I want) is cheap and easy sandboxing for development. This means the following to me: I can modify any package locally, including its dependencies without the fear of breaking anything. That bit about editing depending is the only situation where pins become useful to me. That is to say, relatively rarely.

Opam seems to already give me a tool to accomplish that: the switch. What I'm missing is a convenient UI (or some discipline) to prevent me from accidentally installing things to global switches, not have to remember which switch I need to recall for a particular repository I'm working on, not recall project specific build/test incantation, etc.

I was under the impression that this new `$ opam build` feature gave me something to something along those needs. Am I wrong?

**Comment by dbuenzli** [comment]
> Although I don't think convenience necessarily muddies anything.

It works exactly the converse way say a `make` build system would work: `make install` will build if needed and then install, `make` (i.e. `build`) will only build, never install.

You can see that as a dependency relationship between the fields in `opam` files, for being able to evaluate the `install:` field I first need to evaluate the `build:` field. Now if you install on `build:` you do it the wrong and confusing way round in my opinion.

> I was under the impression that this new $ opam build feature gave me something to something along those needs. Am I wrong?

AFAIU `opam build` completely.

I think that what you want for this is local switches + the ability to interpret the fields of a local `opam` file in this switch, e.g. the `deps:` or `build:` instructions (to be done in the cwd rather than in an opam switch build dir), but `install:` would be dubious, you have pins for this.

**Comment by dbuenzli** [comment]
> AFAIU opam build completely.

Which I characterized somewhere else as this: "A transient pin install that possibly automatically creates a switch with the ability to redirect the result of the install", not something I personally find especially useful.

**Comment by Drup** [comment]
I just want to point out that, no, `opam build` doesn't install in general. Apparently there is an option to do that, which is indeed a bit dubious, and you should redirect your grief towards that specific option.

The command in general is basically "install all the dependencies (which may involve installing ocaml itself) and then execute the build command in the local dev source tree".

**Comment by rgrinberg** [comment]
>Which I characterized somewhere else as this: "A transient pin install that possibly automatically creates a switch with the ability to redirect the result of the install", not something I personally find especially useful.

I can see how that may be the case for you. But consider a user who is mostly a
consumer of opam libraries. I find myself in that spot a lot. In that spot I
very rarely:

* Synchronize multi package opam releases
* Develop multiple packages at once
* Maintain numerous libraries on opam

What I do is mostly maintain up to a dozen of independent applications that
depend on various opam packages. In such a workflow, I pin packages only as a
last resort when I need a particular bug fix or feature. I need to maintain
strict separation between such applications. For example, when I'm upgrading the
dependencies of one application, I must not break how another application
builds. In the rare cases when I make changes to upstream dependencies, those
are usually small, isolated bug fixes that take up to a few days of work. I'd
that workflow of fixing and submitting things upstream to be as simple as
possible, so that I can continue with my own development. I never want to risk
breaking my work switches while doing such side work

Anyway, what I'm trying to say is that this isn't some niche use case. It's
likely to be the majority of opam users who aren't necessarily library authors.

**Comment by lpw25** [comment]
My 2 cents:

> - if opam hasn’t been initialised (no ~/.opam), this is taken care of.
> - if no switch is otherwise explicitely selected, a local switch is used, and
      created if necessary (i.e. in ./_opam/)
> - the metadata for the current project is registered, and the package installed
  after its dependencies, as opam usually does

(from the blog post)

The first of these points seems orthogonal to the rest. If auto-initialising is useful then I would suggest to do it with a prompt ("opam has not been initialised. Initialise it now? [Y/n]") and do it on every command.

The last point does seem to be an install command rather than a build command. I suspect some of the driving force behind making `opam build` is that for some peoples workflow this style of install is more common than the traditional `opam install` and so they want it to have a short name. In which case how about trying to support `opam install .`/`opam install ./opam` instead? So the thing you install is either a package name or a directory/opam file. If you only allow paths that start with `/`, `.` or `..` it shouldn't ever be ambiguous.

The middle point could also be handled in this manner: if the thing being installed is a package name then default to the current switch (local or global), if the thing being installed is a directory then default to a local switch (creating one if it doesn't already exist).

For the other features mentioned in the article:
- `--install-prefix` could just be an option to `install` since it seems like it would work for any installation command
- `--repositories` and any other option for creating local switches should be allowed on `install` in order to customize the local switch that might be created (and maybe error if no such switch was created).

**Comment by AltGr** [comment]
Right, there seems to be some confusion around this...

TL;DR: autoinit stuff put apart, `opam build` behaves  just like pinning, but without registering the pin.

>  if you don't extract the install via --install-dir it will install in your switch and will not recompile if dependents upgrade, hello make inconsistent assumptions about interface messages.

It won't get to that: opam will never keep installed a package that would be inconsistent with changes in the switch universe. The result of `opam build` on `foo`, assuming the switch already exists, would actually be the same as:
```
opam pin . -n
opam install foo
opam unpin foo -n
```

That is, the package is installed and correctly registered, but not pinned to its local source, so that state is not permanent. That means that if a recompilation is required, opam will fall it back to its definition and source from the repositories if one exists. If not and you requested `opam upgrade`, opam is going to prompt you for _removing_ the package.

If that is not what you want, then what you wanted must have been to pin the package in the first place. I experimented with automatic pinning, though, back in september (after the alpha, when we already had the local switches and `opam pin .`, but I was looking for a better interface to wrap that), but it wasn't very nice to work with, which led me to this design. I am glad on any help to improve it yet !

> The command in general is basically "install all the dependencies (which may involve installing ocaml itself) and then execute the build commands in the local dev source tree".

`opam build` runs all the commands (build, install) within the (local) switch, like `opam install`. There is an option for building in-source that is already available to `opam install` as well, `--inplace-build`, but it's orthogonal¹. In its current implementation, `opam build`
- __does__ build and install _within the current/local switch_
- __may__ _copy_ the results of the installation to the given `--install-prefix`, for convenience, but that's really something completely independent that could be in `opam install` (or even run after the fact in another command), made possible by the opam 2.0 packagefile tracking.

Why do this ? The reason is that we want to support source trees containing multiple, interdependent packages. It wouldn't work without installing the local, intermediate packages, and compiling in-source concurrently would break².

> In which case how about trying to support opam install . and opam install ./opam

Good point. I don't know yet how good an end-user interface that would be, but I rather like it, and it's quite consistent with the way switch names have been overloaded to allow directories, for local switches.

> The middle point could also be handled in this manner: if the thing being installed is a package name then default to the current switch (local or global), if the thing being installed is a directory then default to a local switch (creating one if it doesn't already exist).

This I am not sure about, we probably shouldn't link the two features in the UI while they are in fact independent. This could be summarised to `create a local switch if none exists on "opam install <file-or-dir>"`, though, which mimics most closely the current `opam build` design.

> --repositories and any other option for creating local switches should be allowed on install in order to customize the local switch that might be created (and maybe error if no such switch was created).

That would be a whole lot of options added, that are only meaningful in one specific case. Worse still, if you specify e.g. `--repositories`, and the switch already existed, it is not clear what happens. TBH `opam build` already has this problem:

       `--repositories=REPOS`
           When creating a new switch, use the given selection of repositories
           instead of the default. `REPOS` should be a comma-separated list of
           either already registered repository names (configured through e.g.
           `opam repository add --dont-select`), or `NAME=URL` bindings, in which
           case `NAME` should not be registered already to a different URL, and
           the new repository will be registered. See opam repository for more
           details. If the switch already exists, and it is local (`./`), its
           repositories are reset. The argument is otherwise ignored.

The interface would be much clearer with this being an option of `opam switch create` only. While I agree that having to register the repositories beforehand using `opam repository add --dont-select` was quite annoying, is it that important to be able to do anything in a single command ?

¹ Experimenting with `--inplace-build`, I didn't find it convenient at all, because it re-runs `./configure`, with its own parameters, every time. Of course, with more modern build systems, it may be less of a burden.
² `--inplace-build` is still supported in that case, but you'll lose some parallelism.

**Comment by Drup** [comment]
Ok, so apparently I got it wrong, and now I agree with @dbuenzli: the semantics is weird and the name is wrong for what it does. Also, the man page is extremely misleading (emphasis mine):

> opam-build - Work with opam-aware source trees
> This command is expected to be run from the the root of a source tree,
       containing package specifications (one or more opam files). Without
       arguments, it installs the corresponding dependencies and **builds the
       packages locally**, but **without installing or registering them** in the
       opam installation, as a package pinning would do.

What you described above is "do opam install, but only in a temporary way", which is really not what I would expect. 

@rgrinberg, it seems to me that your workflow mostly requires pins+local switches (and I use that exact workflow, but with regular switches). Local switches are the real great addition for your workflow. The "install but not really" part of `opam build` does not seem of any use to you.

In the blog post, @altgr mentions this use case:
```
$ opam build --deps-only
$ eval $(opam config env)
$ ./configure; make; etc.
```
Which I understand, and it seems very useful to me. It also doesn't use the "temporary install" part. Does anyone has an example workflow that would benefit from it?

**Comment by dbuenzli** [comment]
@AltGr Thanks for the clarifications. 

Given that it seems there's agreement nobody obviously understand what `opam build` does and when we do we find this weird. I think the best would be to simply melt its interesting functionality (if any) into `opam install` were it belongs.

I do understand your install idea to support multiple packages within a single source tree, but I somehow I highly doubt its usefulness: it seems to me that usually in this case the build system is able to build all the packages in the dev repo without having the packages installed.

In any case I think we should keep `opam build` to be what @drup used to think it did, i.e. interpret and run the `build:` instructions of a local opam file in the *current* switch (can be local or global) and in the *cwd* (i.e. not in an opam managed build dir).

> Anyway, what I'm trying to say is that this isn't some niche use case. It's
> likely to be the majority of opam users who aren't necessarily library authors.

@rgrinberg I don't know where you got the idea I thought this was a niche use case, it's of course a very important one for developing apps. However I completely fail to see how `opam build` as implemented now helps you for this.

**Comment by AltGr** [comment]
> I do understand your install idea to support multiple packages within a single source tree, but I somehow I highly doubt its usefulness: it seems to me that usually in this case the build system is able to build all the packages in the dev repo without having the packages installed.

Yes, but in that case, you will want to select the rules you run for the build system and let it handle parallelism, you won't be following the rules in the opam files. So e.g. `install --deps-only ./ && make` would be what you want to do.

**Comment by samoht** [comment]
(I haven't read everything on that thread so apologies if I am off topic -- I'll read more later this week)

My primary use-case for requesting `opam build` was to be able to do something like:

```
# I want to contribute to the project
git clone foo.git
cd foo
opam build
[ do some edits ]
opam build
[ make a PR ]
```

Note:
- auto-init: same as what Leo said, should not be something special for this command.
- auto-install (by changing the destination dir): I am not very keen on that. I just want to build stuff, not to install stuff (otherwise I would use `opam install` :p ). It's also a bit weird to change the install dir when you build.
- one of the main use for me are the multi-package repos such as https://github.com/docker/datakit where the build today is very painful (need to pin all packages locally). I agree that this could be solved by a clever build system, but we don't have it (yet).

Re-note: more generally I think it would be useful to be very explicit about the use-case (e.g. the list of commands that the user should write to accomplish a given task).

**Comment by Drup** [comment]
@samoht what about extending  `opam pin` to pin all the packages in a directory and adding all `opam build`'s fancy auto switch ?

You would then do:
```
git clone foo.git
cd foo
opam pin .
[ do some edits ]
opam update -u
[ make a PR ]
```

And you have real pins, and the whole thing seems a lot more consistent.

**Comment by AltGr** [comment]
No need for an extension, `opam pin .` is already there in opam 2.0.
What may be lacking is a way to refer later to these packages though:
```
git clone foo.git
cd foo
# optionally: opam switch create .
opam pin .
# Until now, all good
[ do some edits ]
opam upgrade <need to list locally pinned packages!>
[ make a PR ]
```

`opam update -u` is a good idea and would work, but you'll pay the time of updating your repositories every time. Plus other pinned packages if you aren't using a local switch.

So maybe `opam upgrade .` for `upgrade any packages pinned with target $CWD` is a missing piece for this workflow ?

**Comment by samoht** [comment]
```
opam pin .
[ do some edits ]
opam upgrade
```
Would it be possible to run the same command again? E.g. I just want to build the damn thing, and now I have to remember to `pin` and `upgrade`.

A very similar use-case is when you have a CI, with a local cache (which is or isn't initialised) and then you just want to check if everything compiles, e.g. you don't want to check first if the cache exist to know which command to run.

**Comment by Drup** [comment]
@samoht Well, it also work if you pin again.

**Comment by samoht** [comment]
> In any case I think we should keep opam build to be what @drup used to think it did, i.e. interpret and run the build: instructions of a local opam file in the current switch (can be local or global) and in the cwd (i.e. not in an opam managed build dir).

I think this works for my use case. But it is not clear what that means for multi-package repos. 

> It works exactly the converse way say a make build system would work: make install will build if needed and then install, make (i.e. build) will only build, never install.

What happens for dependencies that you need to `build` the _PWD_ ? Would it make sense to `install` them? In the case of a multi-package repo, that would mean installing all the missing deps, including the one present in _PWD_.

**Comment by AltGr** [comment]
Additionally, if we want to discard the "install like pinned but without pinning" thingy, we could have `<dir>` accepted instead of package arguments of `install`, `reinstall`, `upgrade`, `remove` to:
- apply to the packages pinned to that target
- if there are none, prompt to pin according to the package definitions found there (?)
- check the package definitions and ask to match them with the pins, updating them if necessary ?

`opam pin .` is idempotent, so you won't have to check first if the cache exists.
I'm a bit wary of auto-pinning, which could lead to people not understanding what they had pinned and consequently not understanding errors later on. Of course this would be less of a problem on local switches.

**Comment by AltGr** [comment]
Ok, so here is a proposal for what might be a better design:

- Remove `opam build`.

- Add the possibility to specify opam files or project directories containing opam files to `opam install` (and `upgrade`, `reinstall`, `remove`, for consistency of CLI). The packages will be automatically pinned (if the "install without pinning" workflow above is useful to some, a much clearer option e.g. `--dont-pin` may be added to explicitely make the pinning non-permanent).

- Extend and leverage `opam switch create <dir>`:
  - auto-initialise if necessary
  - automatically pin local package definitions, if any
  - choose the compiler according to what is compatible with the packages
  - possibly, honor a local config file setting up the created switch

This seems consistent as it's basically the intended usage for local switches, reduces friction and the number of commands and seems to better follow what the user would expect from the commands.

It doesn't, however, allow to repeat the same command for further runs, unless we accept that `opam switch create <dir>` synchronises and reinstalls the packages if the setup is already done, which is a possibility but takes it a bit away from what the name implies.

A simple option to disable the auto-pin behaviour of `opam switch create <dir>` should be provided.

**Comment by AltGr** [comment]
One could argue that people using a global switch will use `opam pin .` explicitely rather than `opam install .`, and, consequently, that it would be possible to automatically create a local switch upon `opam install .`.

However, if you have multiple packages in the directory, `opam pin .` does not offer a subsequent `opam reinstall <what was just pinned>` command at the moment, and that would be very convenient, but contradicted by the auto-local-switch above. Having `opam install .` propose to create a local switch only if no package is pinned to the given directory in the current switch would solve that, but becomes too far-fetched IMO.

**Comment by AltGr** [comment]
#2893 has a first prototype of `opam install <dir>`.

**Comment by dbuenzli** [comment]
@AltGr I think I like your latest proposal. If I understand it correctly we entirely drop the idea of letting opam drive your build system in the `cwd` and always work with pins. 

I think it's the only sane way of working if you have multi packages in a single repo. One issue that remains in this scenario is to avoid rebuilding everything from scratch on each `opam upgrade .`, i.e. persist and reuse the opam managed build dir, which IIRC is already solved by `opam pin --incremental .` (I don't remember the exact name). So in this case the worfklow is : 

```
git clone bla && cd bla
opam pin . 
opam upgrade .
[...]
opam upgrade . 
```
which seems quite clear (if people really want a single command `opam upgrade .` could ask for pinning if it detects unpinned elements in `.`)

I think this design is also quite reasonable for the local dev of an app: 

1. I don't really see a problem of having it named in the switch, if your app has an opam file that describes its deps and build instruction you can aswell see it as a package (which may further help for deployment scenarios down the road, `opam bundle` etc.). 

2. If you are not really interested by 1. and prefer to build your app in the current `cwd` with your build system. Then `opam install --deps-only .` is your friend and simply installs whatever you need to build your app in the current switch. In this case the worfklow is 
    ```
    git clone myapp && cd myapp 
    opam install --deps-only . 
    ./build
    ```

(Btw. a good semantics should be found for `opam install --deps-only .` if you have multiple local opam files whose names refer each other, do you automatically pin the smallest set that is needed ? or only install names that are in the `deps:` fields but not mentioned in `.`).

**Comment by AltGr** [comment]
> (Btw. a good semantics should be found for opam install --deps-only . if you have multiple local opam files whose names refer each other, do you automatically pin the smallest set that is needed ? or only install names that are in the deps: fields but not mentioned in .)

The algorithm for `--deps-only` is: (i) resolve the installation of the specified package normally (but assuming they are available) (ii) get the set of packages marked for installation (iii) remove the reverse dependency cones of the initially specified packages from that set.

So `opam install . --deps-only` would give all packages not in `.` that are required. If packages are interleaved (depending on one package in `./`, but required by another one), which honestly shouldn't happen, they are left out.

Thanks for all the feedback on this, that has helped build a much better solution :) (see here for the related announcement: https://github.com/ocaml/platform-blog/pull/45)
